File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Vector2d.java
package org.firstinspires.ftc.teamcode;

// This class used to use vecmath.Vector2d by extending it,
// but since the vecmath package was having trouble integrating into
// the robotics code build environment, vecmath was just removed instead.

public class Vector2d /* extends javax.vecmath.Vector2d */ {
    public double x;
    public double y;

    public void set(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public Vector2d() {
    }
    public Vector2d(double x, double y) {
        set(x, y);
    }

    // Static helper functions

    public static Vector2d construct() {
        return new Vector2d();
    }
    public static Vector2d construct(double x, double y) {
        return new Vector2d(x, y);
    }
    public static Vector2d polar(double r, double theta) {
        return Vector2d.construct(r, 0).rotate(theta);
    }

    // Rotate vector in place
    // Positive angle is counterclockwise
    // Angle measured in radians
    public Vector2d rotate(double angle) {
        x = x*Math.cos(angle) - y*Math.sin(angle);
        y = y*Math.cos(angle) + x*Math.sin(angle);
        return this;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Vector2d.kt
package org.firstinspires.ftc.teamcode

data class Vector2d(val x: Double, val y: Double) {

    // Identity of a vector.
    constructor() : this(0.0, 0.0)

    // Return a new vector that is rotated by an angle.
    // Positive angle is counterclockwise.
    // Angle measured in radians.
    infix fun rotate(angle: Double) = Vector2d(
        x = x * Math.cos(angle) - y * Math.sin(angle),
        y = y * Math.cos(angle) + x * Math.sin(angle),
    )

    companion object {
        // Static helper functions
        // This will probably generate a syntax error, cause I think "this" here
        // refers to the companion object, not the Vector2d constructor.
        TODO("Testing stuff below as described above")
        // This function allows you to get a vector using polar coordinates
        // instead of cartesian coordinates.
        fun polar(r: Double, theta: Double) = this(r, 0.0).rotate(theta)
    }
}
